File: client/lib/data/datasources/local/assessments/operations/create_assessment.dart
import 'package:likha/core/database/db_schema.dart';
import 'package:likha/core/database/local_database.dart';
import 'package:likha/core/errors/exceptions.dart';
import 'package:sqflite_sqlcipher/sqflite.dart';
import 'package:uuid/uuid.dart';

Future<String> createAssessment(
  LocalDatabase localDatabase,
  String classId,
  String title,
  String? description,
  int timeLimitMinutes,
  String openAt,
  String closeAt,
  bool? showResultsImmediately,
  bool isPublished,
  String? tosId,
  int? gradingPeriodNumber,
  String? component, {
  String? id,
  Transaction? txn,
}) async {
  try {
    final db = await localDatabase.database;
    final now = DateTime.now();
    final assessmentId = id ?? const Uuid().v4();
    final map = {
      CommonCols.id: assessmentId,
      AssessmentsCols.classId: classId,
      AssessmentsCols.title: title,
      AssessmentsCols.description: description,
      AssessmentsCols.timeLimitMinutes: timeLimitMinutes,
      AssessmentsCols.openAt: openAt,
      AssessmentsCols.closeAt: closeAt,
      AssessmentsCols.showResultsImmediately: showResultsImmediately == true ? 1 : 0,
      AssessmentsCols.resultsReleased: 0,
      AssessmentsCols.isPublished: isPublished ? 1 : 0,
      AssessmentsCols.orderIndex: 0,
      if (tosId != null) AssessmentsCols.tosId: tosId,
      if (gradingPeriodNumber != null) AssessmentsCols.gradingPeriodNumber: gradingPeriodNumber,
      if (component != null) AssessmentsCols.component: component,
      CommonCols.createdAt: now.toIso8601String(),
      CommonCols.updatedAt: now.toIso8601String(),
      CommonCols.cachedAt: now.toIso8601String(),
      CommonCols.syncStatus: 'pending',
    };

    if (txn != null) {
      await txn.insert(DbTables.assessments, map);
    } else {
      await db.insert(DbTables.assessments, map);
    }

    return assessmentId;
  } catch (e) {
    throw CacheException('Failed to create assessment locally: $e');
  }
}

File: server/src/modules/assessment/repository_operations/create_assessment.rs
use chrono::Utc;
use sea_orm::*;
use uuid::Uuid;

use ::entity::assessments;
use crate::utils::{AppError, AppResult};

pub async fn create_assessment(
    db: &DatabaseConnection,
    class_id: Uuid,
    title: String,
    description: Option<String>,
    time_limit_minutes: i32,
    open_at: chrono::NaiveDateTime,
    close_at: chrono::NaiveDateTime,
    show_results_immediately: bool,
    order_index: i32,
    client_id: Option<Uuid>,
    is_published: bool,
    term_number: Option<i32>,
    component: Option<String>,
    tos_id: Option<String>,
) -> AppResult<assessments::Model> {
    let tos_id = match tos_id {
        Some(s) if !s.is_empty() => Some(
            Uuid::parse_str(&s)
                .map_err(|e| AppError::BadRequest(format!("Invalid tos_id UUID: {}", e)))?,
        ),
        _ => None,
    };
    let assessment = assessments::ActiveModel {
        id: Set(client_id.unwrap_or_else(Uuid::new_v4)),
        class_id: Set(class_id),
        title: Set(title),
        description: Set(description),
        time_limit_minutes: Set(time_limit_minutes),
        open_at: Set(open_at),
        close_at: Set(close_at),
        show_results_immediately: Set(show_results_immediately),
        results_released: Set(false),
        is_published: Set(is_published),
        order_index: Set(order_index),
        total_points: Set(0),
        created_at: Set(Utc::now().naive_utc()),
        updated_at: Set(Utc::now().naive_utc()),
        deleted_at: Set(None),
        term_number: Set(term_number),
        component: Set(component),
        tos_id: Set(tos_id),
    };

    assessment
        .insert(db)
        .await
        .map_err(|e| AppError::InternalServerError(format!("Failed to create assessment: {}", e)))
}
